python/cuvs/cuvs/neighbors/hnsw/hnsw.pyx

```diff
@@ -0,0 +1,371 @@
+
+    def __cinit__(self):
+        self.trained = False
+        check_cuvs(cuvsHnswIndexCreate(&self.index))
+
+    def __dealloc__(self):
+        if self.index is not NULL:
+            check_cuvs(cuvsHnswIndexDestroy(self.index))
+
+    @property
+    def trained(self):
+        return self.trained
+
+    def __repr__(self):
+        # todo(dgd): update repr as we expose data through C API
+        attr_str = []
+        return "Index(type=HNSW, metric=L2" + (", ".join(attr_str)) + ")"
+
+
+@auto_sync_resources
+def save(filename, cagra.Index index, resources=None):
+    """
+    Saves the CAGRA index to a file as an hnswlib index.
+    The saved index is immutable and can only be searched by the hnswlib
+    wrapper in cuVS, as the format is not compatible with the original
+    hnswlib.
+
+    Saving / loading the index is experimental. The serialization format is
+    subject to change.
+
+    Parameters
+    ----------
+    filename : string
+        Name of the file.
+    index : Index
+        Trained CAGRA index.
+    {resources_docstring}
+
+    Examples
+    --------
+    >>> import cupy as cp
+    >>> from cuvs.neighbors import cagra
+    >>> n_samples = 50000
+    >>> n_features = 50
+    >>> dataset = cp.random.random_sample((n_samples, n_features),
+    ...                                   dtype=cp.float32)
+    >>> # Build index
+    >>> index = cagra.build(cagra.IndexParams(), dataset)
+    >>> # Serialize and deserialize the cagra index built
+    >>> hnsw.save("my_index.bin", index)
+    """
+    cdef string c_filename = filename.encode('utf-8')
+    cdef cuvsResources_t res = <cuvsResources_t>resources.get_c_obj()
+    check_cuvs(cagra.cuvsCagraSerializeToHnswlib(res,
+                                                 c_filename.c_str(),
+                                                 index.index))
+
+
+@auto_sync_resources
+def load(filename, dim, dtype, metric="sqeuclidean", resources=None):
+    """
+    Loads base-layer-only hnswlib index from file, which was originally
+    saved as a built CAGRA index. The loaded index is immutable and can only
+    be searched by the hnswlib wrapper in cuVS, as the format is not
+    compatible with the original hnswlib.
+
+    Saving / loading the index is experimental. The serialization format is
+    subject to change, therefore loading an index saved with a previous
+    version of cuVS is not guaranteed to work.
+
+    Parameters
+    ----------
+    filename : string
+        Name of the file.
+    dim : int
+        Dimensions of the training dataest
+    dtype : np.dtype of the saved index
+        Valid values for dtype: [np.float32, np.byte, np.ubyte]
+    metric : string denoting the metric type, default="sqeuclidean"
+        Valid values for metric: ["sqeuclidean"], where
+            - sqeuclidean is the euclidean distance without the square root
+              operation, i.e.: distance(a,b) = \\sum_i (a_i - b_i)^2,
+            - inner product distance is defined as
```

Comment:
Perhaps this should be worded so that the actual required string is in the sentence like the sqeuclidean? 
```suggestion
            - inner_product distance is defined as
```